pubsub.go

```diff
@@ -186,6 +190,13 @@ func WithValidateThrottle(n int) Option {
 	}
 }
 
+func WithMessageSigning() Option {
```

Comment:
Maybe we should consider using different keys than our nodes peer key?

Reply:
Sure, we can do that too -- the `from` in published messages will have to change to match the signing key though.

Reply:
I think we can have a `WithMessageSigningKey` option that explicitly specifies the signing key.

Reply:
Yeah, `from` can be any valid key. However, I'd prefer it if it were a peer ID (by default at least).

Reply:
Also, this can't be optional.

Reply:
Sure, ID works -- we can pull the key from the peerstore anyway.

> Also, this can't be optional.

Do you want to specify the ID in the option instead of relying on `host.ID()` for the default and using a separate option?

Reply:
We can always vararg an optional peer ID argument in the option.
If it is not included, use the host ID; otherwise use the specified ID to retrieve the key from the peerstore.

Reply:
> Also, this can't be optional.

@Stebalien did you mean you want message signing to be the default?

Reply:
Given that we deduplicate messages based on `(from, seqno)`, message signing must be mandatory.

Reply:
How about we have a `strict` flag as an argument to `WithMessageSigning`?
When true, then reject unsigned messages. No default, the user gets to specify behaviour.

Reply:
We shouldn't *enforce* message signing by default, for now, unless we're willing to break things. However, we should *always* sign messages (or switch to deduplicating based on the hash).

Reply:
Signing without enforcing the signatures doesn't add anything over not siging wrt to message ID collision attacks. 
I think we should have the strict option when enabling it, but keep it disabled by default (at least in the near term).

Reply:
> Signing without enforcing the signatures doesn't add anything over not siging wrt to message ID collision attacks.

Sorry, I was being lazy instead of explicit.

IMO:

1. We should *always* sign.
2. For *now*, we should probably disable verification by default for backwards compatibility (unless we're willing to break things).
3. We should definitely provide some form of strict option, defaulting to false. We'll eventually switch it to true by default and then, make it a no-op (someday).

Reply:
1. I don't know about this, it has perfomance implications.
2. that's the approach here, so good.
3. we can just add the strict option to `WithMessageSigning` which lets the user in the loop.  we can make it have a default value by making it a variadic interface.

Reply:
@whyrusleeping thoughts on the "always sign" matter?

Reply:
Also, we don't have to make it "awlays sign" in the first step; let's add the signatures, have some usage to see how it fares in practice perf wise, and then we can decide whether to always sign in a later iteration.
I think it's a bit too drastic to go "always sign" in one step.

Reply:
> I don't know about this, it has perfomance implications.

I'd *rather* use CIDs or hashes to identify messages but we need to do something.

Reply:
Well, ideally we would use message signing with strict mode.
I would argue that it's sane to keep the default behaviour unchanged (for a while at least), and provide the easy upgrade path of `WithMessageSigning(true)` for applications that need to secure their pubsub.
Eventually we can make it the default.

I don't think we need to take reflexive action and dump our simple message id construction just yet -- note that using hashes also has _some_ perf implications.

Reply:
> I don't think we need to take reflexive action

Either is fine by me. My preference is to layer things (handle authorship at a higher layer) but it *is* useful to have that baked into the bottom layer.

---

> I would argue that it's sane to keep the default behaviour unchanged (for a while at least), and provide the easy upgrade path of `WithMessageSigning(true)` for applications that need to secure their pubsub.

I'm not worried about "security", I'm worried about griefing by bored children (e.g., what happens on IRC). However, as long as we turn it on by default in go-ipfs (with strict off), I can live with it.

Reply:
Makes sense to enable it in go-ipfs, I'll add the relevant options when we integrate it.